--- a/net/minecraft/world/item/crafting/BlastingRecipe.java
+++ b/net/minecraft/world/item/crafting/BlastingRecipe.java
@@ -4,7 +4,16 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.block.Blocks;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.v1_19_R1.inventory.CraftBlastingRecipe;
+import org.bukkit.craftbukkit.v1_19_R1.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.v1_19_R1.inventory.CraftRecipe;
+import org.bukkit.craftbukkit.v1_19_R1.util.CraftNamespacedKey;
+import org.bukkit.inventory.Recipe;
+// CraftBukkit end
+
 public class BlastingRecipe extends AbstractCookingRecipe {
+
     public BlastingRecipe(ResourceLocation id, String group, Ingredient input, ItemStack output, float experience, int cookTime) {
         super(RecipeType.BLASTING, id, group, input, output, experience, cookTime);
     }
@@ -18,4 +27,16 @@
     public RecipeSerializer<?> getSerializer() {
         return RecipeSerializer.BLASTING_RECIPE;
     }
+
+    // CraftBukkit start
+    @Override
+    public Recipe toBukkitRecipe() {
+        CraftItemStack result = CraftItemStack.asCraftMirror(this.result);
+
+        CraftBlastingRecipe recipe = new CraftBlastingRecipe(CraftNamespacedKey.fromMinecraft(this.id), result, CraftRecipe.toBukkit(this.ingredient), this.experience, this.cookingTime);
+        recipe.setGroup(this.group);
+
+        return recipe;
+    }
+    // CraftBukkit end
 }
