--- a/net/minecraft/world/item/crafting/RecipeManager.java
+++ b/net/minecraft/world/item/crafting/RecipeManager.java
@@ -1,6 +1,7 @@
 package net.minecraft.world.item.crafting;
 
 import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableMap.Builder;
 import com.google.common.collect.Maps;
 import com.google.gson.Gson;
 import com.google.gson.GsonBuilder;
@@ -13,8 +14,10 @@
 import java.util.Collection;
 import java.util.Collections;
 import java.util.Comparator;
+import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
+import java.util.Map.Entry;
 import java.util.Optional;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
@@ -31,67 +34,99 @@
 import net.minecraft.world.level.Level;
 import org.slf4j.Logger;
 
+import it.unimi.dsi.fastutil.objects.Object2ObjectLinkedOpenHashMap; // CraftBukkit
+
 public class RecipeManager extends SimpleJsonResourceReloadListener {
+
     private static final Gson GSON = (new GsonBuilder()).setPrettyPrinting().disableHtmlEscaping().create();
     private static final Logger LOGGER = LogUtils.getLogger();
-    public Map<RecipeType<?>, Map<ResourceLocation, Recipe<?>>> recipes = ImmutableMap.of();
+    public Map<RecipeType<?>, Object2ObjectLinkedOpenHashMap<ResourceLocation, Recipe<?>>> recipes = ImmutableMap.of(); // CraftBukkit
     public Map<ResourceLocation, Recipe<?>> byName = ImmutableMap.of();
     private boolean hasErrors;
 
     public RecipeManager() {
-        super(GSON, "recipes");
+        super(RecipeManager.GSON, "recipes");
     }
 
-    @Override
     protected void apply(Map<ResourceLocation, JsonElement> prepared, ResourceManager manager, ProfilerFiller profiler) {
         this.hasErrors = false;
-        Map<RecipeType<?>, ImmutableMap.Builder<ResourceLocation, Recipe<?>>> map = Maps.newHashMap();
-        ImmutableMap.Builder<ResourceLocation, Recipe<?>> builder = ImmutableMap.builder();
+        // CraftBukkit start - SPIGOT-5667 make sure all types are populated and mutable
+        Map<RecipeType<?>, Object2ObjectLinkedOpenHashMap<ResourceLocation, Recipe<?>>> map1 = Maps.newHashMap();
+        for (RecipeType<?> recipeType : Registry.RECIPE_TYPE) {
+            map1.put(recipeType, new Object2ObjectLinkedOpenHashMap<>());
+        }
+        // CraftBukkit end
+        Builder<ResourceLocation, Recipe<?>> builder = ImmutableMap.builder();
+        Iterator iterator = prepared.entrySet().iterator();
 
-        for(Map.Entry<ResourceLocation, JsonElement> entry : prepared.entrySet()) {
-            ResourceLocation resourceLocation = entry.getKey();
+        while (iterator.hasNext()) {
+            Entry<ResourceLocation, JsonElement> entry = (Entry) iterator.next();
+            ResourceLocation minecraftkey = (ResourceLocation) entry.getKey();
 
             try {
-                Recipe<?> recipe = fromJson(resourceLocation, GsonHelper.convertToJsonObject(entry.getValue(), "top element"));
-                map.computeIfAbsent(recipe.getType(), (recipeType) -> {
-                    return ImmutableMap.builder();
-                }).put(resourceLocation, recipe);
-                builder.put(resourceLocation, recipe);
-            } catch (IllegalArgumentException | JsonParseException var10) {
-                LOGGER.error("Parsing error loading recipe {}", resourceLocation, var10);
+                Recipe<?> irecipe = RecipeManager.fromJson(minecraftkey, GsonHelper.convertToJsonObject((JsonElement) entry.getValue(), "top element"));
+
+                // CraftBukkit start
+                (map1.computeIfAbsent(irecipe.getType(), (recipes) -> {
+                    return new Object2ObjectLinkedOpenHashMap<>();
+                    // CraftBukkit end
+                })).put(minecraftkey, irecipe);
+                builder.put(minecraftkey, irecipe);
+            } catch (IllegalArgumentException | JsonParseException jsonparseexception) {
+                RecipeManager.LOGGER.error("Parsing error loading recipe {}", minecraftkey, jsonparseexception);
             }
         }
 
-        this.recipes = map.entrySet().stream().collect(ImmutableMap.toImmutableMap(Map.Entry::getKey, (entryx) -> {
-            return entryx.getValue().build();
+        this.recipes = (Map) map1.entrySet().stream().collect(ImmutableMap.toImmutableMap(Entry::getKey, (entry1) -> {
+            return entry1.getValue(); // CraftBukkit // Paper - decompile fix - *shrugs internally* // todo: is this needed anymore?
         }));
-        this.byName = builder.build();
-        LOGGER.info("Loaded {} recipes", (int)map.size());
+        this.byName = Maps.newHashMap(builder.build()); // CraftBukkit
+        RecipeManager.LOGGER.info("Loaded {} recipes", map1.size());
+    }
+
+    // CraftBukkit start
+    public void addRecipe(Recipe<?> irecipe) {
+        org.spigotmc.AsyncCatcher.catchOp("Recipe Add"); // Spigot
+        Object2ObjectLinkedOpenHashMap<ResourceLocation, Recipe<?>> map = this.recipes.get(irecipe.getType()); // CraftBukkit
+
+        if (this.byName.containsKey(irecipe.getId()) || map.containsKey(irecipe.getId())) {
+            throw new IllegalStateException("Duplicate recipe ignored with ID " + irecipe.getId());
+        } else {
+            map.putAndMoveToFirst(irecipe.getId(), irecipe); // CraftBukkit - SPIGOT-4638: last recipe gets priority
+            this.byName.put(irecipe.getId(), irecipe);
+        }
     }
+    // CraftBukkit end
 
     public boolean hadErrorsLoading() {
         return this.hasErrors;
     }
 
     public <C extends Container, T extends Recipe<C>> Optional<T> getRecipeFor(RecipeType<T> type, C inventory, Level world) {
-        return this.byType(type).values().stream().filter((recipe) -> {
-            return recipe.matches(inventory, world);
+        // CraftBukkit start
+        Optional<T> recipe = this.byType(type).values().stream().filter((irecipe) -> {
+            return irecipe.matches(inventory, world);
         }).findFirst();
+        inventory.setCurrentRecipe(recipe.orElse(null)); // CraftBukkit - Clear recipe when no recipe is found
+        // CraftBukkit end
+        return recipe;
     }
 
     public <C extends Container, T extends Recipe<C>> Optional<Pair<ResourceLocation, T>> getRecipeFor(RecipeType<T> type, C inventory, Level world, @Nullable ResourceLocation id) {
         Map<ResourceLocation, T> map = this.byType(type);
+
         if (id != null) {
-            T recipe = map.get(id);
-            if (recipe != null && recipe.matches(inventory, world)) {
-                return Optional.of(Pair.of(id, recipe));
+            T t0 = map.get(id); // CraftBukkit - decompile error
+
+            if (t0 != null && t0.matches(inventory, world)) {
+                return Optional.of(Pair.of(id, t0));
             }
         }
 
         return map.entrySet().stream().filter((entry) -> {
-            return entry.getValue().matches(inventory, world);
+            return ((Recipe) entry.getValue()).matches(inventory, world);
         }).findFirst().map((entry) -> {
-            return Pair.of(entry.getKey(), entry.getValue());
+            return Pair.of((ResourceLocation) entry.getKey(), entry.getValue()); // CraftBukkit - decompile error
         });
     }
 
@@ -100,38 +135,39 @@
     }
 
     public <C extends Container, T extends Recipe<C>> List<T> getRecipesFor(RecipeType<T> type, C inventory, Level world) {
-        return this.byType(type).values().stream().filter((recipe) -> {
-            return recipe.matches(inventory, world);
-        }).sorted(Comparator.comparing((recipe) -> {
-            return recipe.getResultItem().getDescriptionId();
+        return (List) this.byType(type).values().stream().filter((irecipe) -> {
+            return irecipe.matches(inventory, world);
+        }).sorted(Comparator.comparing((irecipe) -> {
+            return irecipe.getResultItem().getDescriptionId();
         })).collect(Collectors.toList());
     }
 
     private <C extends Container, T extends Recipe<C>> Map<ResourceLocation, T> byType(RecipeType<T> type) {
-        return this.recipes.getOrDefault(type, Collections.emptyMap());
+        return (Map) this.recipes.getOrDefault(type, new Object2ObjectLinkedOpenHashMap<>()); // CraftBukkit
     }
 
     public <C extends Container, T extends Recipe<C>> NonNullList<ItemStack> getRemainingItemsFor(RecipeType<T> type, C inventory, Level world) {
         Optional<T> optional = this.getRecipeFor(type, inventory, world);
+
         if (optional.isPresent()) {
-            return optional.get().getRemainingItems(inventory);
+            return ((Recipe) optional.get()).getRemainingItems(inventory);
         } else {
-            NonNullList<ItemStack> nonNullList = NonNullList.withSize(inventory.getContainerSize(), ItemStack.EMPTY);
+            NonNullList<ItemStack> nonnulllist = NonNullList.withSize(inventory.getContainerSize(), ItemStack.EMPTY);
 
-            for(int i = 0; i < nonNullList.size(); ++i) {
-                nonNullList.set(i, inventory.getItem(i));
+            for (int i = 0; i < nonnulllist.size(); ++i) {
+                nonnulllist.set(i, inventory.getItem(i));
             }
 
-            return nonNullList;
+            return nonnulllist;
         }
     }
 
     public Optional<? extends Recipe<?>> byKey(ResourceLocation id) {
-        return Optional.ofNullable(this.byName.get(id));
+        return Optional.ofNullable(this.byName.get(id)); // CraftBukkit - decompile error
     }
 
     public Collection<Recipe<?>> getRecipes() {
-        return this.recipes.values().stream().flatMap((map) -> {
+        return (Collection) this.recipes.values().stream().flatMap((map) -> {
             return map.values().stream();
         }).collect(Collectors.toSet());
     }
@@ -143,30 +179,53 @@
     }
 
     public static Recipe<?> fromJson(ResourceLocation id, JsonObject json) {
-        String string = GsonHelper.getAsString(json, "type");
-        return Registry.RECIPE_SERIALIZER.getOptional(new ResourceLocation(string)).orElseThrow(() -> {
-            return new JsonSyntaxException("Invalid or unsupported recipe type '" + string + "'");
-        }).fromJson(id, json);
+        String s = GsonHelper.getAsString(json, "type");
+
+        return ((RecipeSerializer) Registry.RECIPE_SERIALIZER.getOptional(new ResourceLocation(s)).orElseThrow(() -> {
+            return new JsonSyntaxException("Invalid or unsupported recipe type '" + s + "'");
+        })).fromJson(id, json);
     }
 
     public void replaceRecipes(Iterable<Recipe<?>> recipes) {
         this.hasErrors = false;
-        Map<RecipeType<?>, Map<ResourceLocation, Recipe<?>>> map = Maps.newHashMap();
-        ImmutableMap.Builder<ResourceLocation, Recipe<?>> builder = ImmutableMap.builder();
-        recipes.forEach((recipe) -> {
-            Map<ResourceLocation, Recipe<?>> map2 = map.computeIfAbsent(recipe.getType(), (t) -> {
-                return Maps.newHashMap();
+        Map<RecipeType<?>, Object2ObjectLinkedOpenHashMap<ResourceLocation, Recipe<?>>> map = Maps.newHashMap(); // CraftBukkit
+        Builder<ResourceLocation, Recipe<?>> builder = ImmutableMap.builder();
+
+        recipes.forEach((irecipe) -> {
+            Map<ResourceLocation, Recipe<?>> map1 = (Map) map.computeIfAbsent(irecipe.getType(), (recipes_) -> { // Paper - remap fix
+                return new Object2ObjectLinkedOpenHashMap<>(); // CraftBukkit
             });
-            ResourceLocation resourceLocation = recipe.getId();
-            Recipe<?> recipe2 = map2.put(resourceLocation, recipe);
-            builder.put(resourceLocation, recipe);
-            if (recipe2 != null) {
-                throw new IllegalStateException("Duplicate recipe ignored with ID " + resourceLocation);
+            ResourceLocation minecraftkey = irecipe.getId();
+            Recipe<?> irecipe1 = (Recipe) map1.put(minecraftkey, irecipe);
+
+            builder.put(minecraftkey, irecipe);
+            if (irecipe1 != null) {
+                throw new IllegalStateException("Duplicate recipe ignored with ID " + minecraftkey);
             }
         });
         this.recipes = ImmutableMap.copyOf(map);
-        this.byName = builder.build();
+        this.byName = Maps.newHashMap(builder.build()); // CraftBukkit
+    }
+
+    // CraftBukkit start
+    public boolean removeRecipe(ResourceLocation mcKey) {
+        for (Object2ObjectLinkedOpenHashMap<ResourceLocation, Recipe<?>> recipes : this.recipes.values()) {
+            recipes.remove(mcKey);
+        }
+
+        return this.byName.remove(mcKey) != null;
+    }
+
+    public void clearRecipes() {
+        this.recipes = Maps.newHashMap();
+
+        for (RecipeType<?> recipeType : Registry.RECIPE_TYPE) {
+            this.recipes.put(recipeType, new Object2ObjectLinkedOpenHashMap<>());
+        }
+
+        this.byName = Maps.newHashMap();
     }
+    // CraftBukkit end
 
     public static <C extends Container, T extends Recipe<C>> RecipeManager.CachedCheck<C, T> createCheck(final RecipeType<T> type) {
         return new RecipeManager.CachedCheck<C, T>() {
@@ -175,12 +234,14 @@
 
             @Override
             public Optional<T> getRecipeFor(C inventory, Level world) {
-                RecipeManager recipeManager = world.getRecipeManager();
-                Optional<Pair<ResourceLocation, T>> optional = recipeManager.getRecipeFor(type, inventory, world, this.lastRecipe);
+                RecipeManager craftingmanager = world.getRecipeManager();
+                Optional<Pair<ResourceLocation, T>> optional = craftingmanager.getRecipeFor(type, inventory, world, this.lastRecipe);
+
                 if (optional.isPresent()) {
-                    Pair<ResourceLocation, T> pair = optional.get();
-                    this.lastRecipe = pair.getFirst();
-                    return Optional.of(pair.getSecond());
+                    Pair<ResourceLocation, T> pair = (Pair) optional.get();
+
+                    this.lastRecipe = (ResourceLocation) pair.getFirst();
+                    return Optional.of(pair.getSecond()); // CraftBukkit - decompile error
                 } else {
                     return Optional.empty();
                 }
@@ -189,6 +250,7 @@
     }
 
     public interface CachedCheck<C extends Container, T extends Recipe<C>> {
+
         Optional<T> getRecipeFor(C inventory, Level world);
     }
 }
