--- a/net/minecraft/world/level/storage/DimensionDataStorage.java
+++ b/net/minecraft/world/level/storage/DimensionDataStorage.java
@@ -120,6 +120,7 @@
 
             pushbackInputStream.close();
         } catch (Throwable var15) {
+            com.destroystokyo.paper.exception.ServerInternalException.reportInternalException(var15); // Paper
             try {
                 fileInputStream.close();
             } catch (Throwable var10) {
