--- a/net/minecraft/world/level/levelgen/feature/treedecorators/CocoaDecorator.java
+++ b/net/minecraft/world/level/levelgen/feature/treedecorators/CocoaDecorator.java
@@ -25,6 +25,7 @@
 
     @Override
     public void place(TreeDecorator.Context generator) {
+        if (generator.logs().isEmpty()) return; // Paper
         RandomSource randomSource = generator.random();
         if (!(randomSource.nextFloat() >= this.probability)) {
             List<BlockPos> list = generator.logs();
