--- a/net/minecraft/world/item/LeadItem.java
+++ b/net/minecraft/world/item/LeadItem.java
@@ -1,8 +1,11 @@
 package net.minecraft.world.item;
 
+import java.util.Iterator;
+import java.util.List;
 import net.minecraft.core.BlockPos;
 import net.minecraft.tags.BlockTags;
 import net.minecraft.world.InteractionResult;
+import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.Mob;
 import net.minecraft.world.entity.decoration.LeashFenceKnotEntity;
 import net.minecraft.world.entity.player.Player;
@@ -11,50 +14,84 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.phys.AABB;
+// CraftBukkit start
+import org.bukkit.craftbukkit.v1_19_R1.CraftEquipmentSlot;
+import org.bukkit.event.hanging.HangingPlaceEvent;
+// CraftBukkit end
 
 public class LeadItem extends Item {
+
     public LeadItem(Item.Properties settings) {
         super(settings);
     }
 
     @Override
     public InteractionResult useOn(UseOnContext context) {
-        Level level = context.getLevel();
-        BlockPos blockPos = context.getClickedPos();
-        BlockState blockState = level.getBlockState(blockPos);
-        if (blockState.is(BlockTags.FENCES)) {
-            Player player = context.getPlayer();
-            if (!level.isClientSide && player != null) {
-                bindPlayerMobs(player, level, blockPos);
+        Level world = context.getLevel();
+        BlockPos blockposition = context.getClickedPos();
+        BlockState iblockdata = world.getBlockState(blockposition);
+
+        if (iblockdata.is(BlockTags.FENCES)) {
+            Player entityhuman = context.getPlayer();
+
+            if (!world.isClientSide && entityhuman != null) {
+                LeadItem.bindPlayerMobs(entityhuman, world, blockposition, context.getHand()); // CraftBukkit - Pass hand
             }
 
-            level.gameEvent(GameEvent.BLOCK_ATTACH, blockPos, GameEvent.Context.of(player));
-            return InteractionResult.sidedSuccess(level.isClientSide);
+            world.gameEvent(GameEvent.BLOCK_ATTACH, blockposition, GameEvent.Context.of((Entity) entityhuman));
+            return InteractionResult.sidedSuccess(world.isClientSide);
         } else {
             return InteractionResult.PASS;
         }
     }
 
-    public static InteractionResult bindPlayerMobs(Player player, Level world, BlockPos pos) {
-        LeashFenceKnotEntity leashFenceKnotEntity = null;
-        boolean bl = false;
-        double d = 7.0D;
-        int i = pos.getX();
-        int j = pos.getY();
-        int k = pos.getZ();
+    public static InteractionResult bindPlayerMobs(Player entityhuman, Level world, BlockPos blockposition, net.minecraft.world.InteractionHand enumhand) { // CraftBukkit - Add EnumHand
+        LeashFenceKnotEntity entityleash = null;
+        boolean flag = false;
+        double d0 = 7.0D;
+        int i = blockposition.getX();
+        int j = blockposition.getY();
+        int k = blockposition.getZ();
+        List<Mob> list = world.getEntitiesOfClass(Mob.class, new AABB((double) i - 7.0D, (double) j - 7.0D, (double) k - 7.0D, (double) i + 7.0D, (double) j + 7.0D, (double) k + 7.0D));
+        Iterator iterator = list.iterator();
 
-        for(Mob mob : world.getEntitiesOfClass(Mob.class, new AABB((double)i - 7.0D, (double)j - 7.0D, (double)k - 7.0D, (double)i + 7.0D, (double)j + 7.0D, (double)k + 7.0D))) {
-            if (mob.getLeashHolder() == player) {
-                if (leashFenceKnotEntity == null) {
-                    leashFenceKnotEntity = LeashFenceKnotEntity.getOrCreateKnot(world, pos);
-                    leashFenceKnotEntity.playPlacementSound();
+        while (iterator.hasNext()) {
+            Mob entityinsentient = (Mob) iterator.next();
+
+            if (entityinsentient.getLeashHolder() == entityhuman) {
+                if (entityleash == null) {
+                    entityleash = LeashFenceKnotEntity.getOrCreateKnot(world, blockposition);
+
+                    // CraftBukkit start - fire HangingPlaceEvent
+                    org.bukkit.inventory.EquipmentSlot hand = CraftEquipmentSlot.getHand(enumhand);
+                    HangingPlaceEvent event = new HangingPlaceEvent((org.bukkit.entity.Hanging) entityleash.getBukkitEntity(), entityhuman != null ? (org.bukkit.entity.Player) entityhuman.getBukkitEntity() : null, world.getWorld().getBlockAt(i, j, k), org.bukkit.block.BlockFace.SELF, hand);
+                    world.getCraftServer().getPluginManager().callEvent(event);
+
+                    if (event.isCancelled()) {
+                        entityleash.discard();
+                        return InteractionResult.PASS;
+                    }
+                    // CraftBukkit end
+                    entityleash.playPlacementSound();
                 }
 
-                mob.setLeashedTo(leashFenceKnotEntity, true);
-                bl = true;
+                // CraftBukkit start
+                if (org.bukkit.craftbukkit.v1_19_R1.event.CraftEventFactory.callPlayerLeashEntityEvent(entityinsentient, entityleash, entityhuman, enumhand).isCancelled()) {
+                    continue;
+                }
+                // CraftBukkit end
+
+                entityinsentient.setLeashedTo(entityleash, true);
+                flag = true;
             }
         }
 
-        return bl ? InteractionResult.SUCCESS : InteractionResult.PASS;
+        return flag ? InteractionResult.SUCCESS : InteractionResult.PASS;
+    }
+
+    // CraftBukkit start
+    public static InteractionResult bindPlayerMobs(Player player, Level world, BlockPos pos) {
+        return LeadItem.bindPlayerMobs(player, world, pos, net.minecraft.world.InteractionHand.MAIN_HAND);
     }
+    // CraftBukkit end
 }
