--- a/net/minecraft/nbt/ByteArrayTag.java
+++ b/net/minecraft/nbt/ByteArrayTag.java
@@ -1,3 +1,4 @@
+// mc-dev import
 package net.minecraft.nbt;
 
 import java.io.DataInput;
@@ -8,24 +9,29 @@
 import org.apache.commons.lang3.ArrayUtils;
 
 public class ByteArrayTag extends CollectionTag<ByteTag> {
+
     private static final int SELF_SIZE_IN_BITS = 192;
     public static final TagType<ByteArrayTag> TYPE = new TagType.VariableSize<ByteArrayTag>() {
         @Override
-        public ByteArrayTag load(DataInput dataInput, int i, NbtAccounter nbtAccounter) throws IOException {
-            nbtAccounter.accountBits(192L);
-            int j = dataInput.readInt();
-            nbtAccounter.accountBits(8L * (long)j);
-            byte[] bs = new byte[j];
-            dataInput.readFully(bs);
-            return new ByteArrayTag(bs);
+        public ByteArrayTag load(DataInput input, int depth, NbtAccounter tracker) throws IOException {
+            tracker.accountBits(192L);
+            int j = input.readInt();
+            com.google.common.base.Preconditions.checkArgument( j < 1 << 24); // Spigot
+
+            tracker.accountBits(8L * (long) j);
+            byte[] abyte = new byte[j];
+
+            input.readFully(abyte);
+            return new ByteArrayTag(abyte);
         }
 
         @Override
         public StreamTagVisitor.ValueResult parse(DataInput input, StreamTagVisitor visitor) throws IOException {
             int i = input.readInt();
-            byte[] bs = new byte[i];
-            input.readFully(bs);
-            return visitor.visit(bs);
+            byte[] abyte = new byte[i];
+
+            input.readFully(abyte);
+            return visitor.visit(abyte);
         }
 
         @Override
@@ -50,18 +56,19 @@
     }
 
     public ByteArrayTag(List<Byte> value) {
-        this(toArray(value));
+        this(ByteArrayTag.toArray(value));
     }
 
     private static byte[] toArray(List<Byte> list) {
-        byte[] bs = new byte[list.size()];
+        byte[] abyte = new byte[list.size()];
 
-        for(int i = 0; i < list.size(); ++i) {
-            Byte byte_ = list.get(i);
-            bs[i] = byte_ == null ? 0 : byte_;
+        for (int i = 0; i < list.size(); ++i) {
+            Byte obyte = (Byte) list.get(i);
+
+            abyte[i] = obyte == null ? 0 : obyte;
         }
 
-        return bs;
+        return abyte;
     }
 
     @Override
@@ -77,7 +84,7 @@
 
     @Override
     public TagType<ByteArrayTag> getType() {
-        return TYPE;
+        return ByteArrayTag.TYPE;
     }
 
     @Override
@@ -87,21 +94,16 @@
 
     @Override
     public Tag copy() {
-        byte[] bs = new byte[this.data.length];
-        System.arraycopy(this.data, 0, bs, 0, this.data.length);
-        return new ByteArrayTag(bs);
+        byte[] abyte = new byte[this.data.length];
+
+        System.arraycopy(this.data, 0, abyte, 0, this.data.length);
+        return new ByteArrayTag(abyte);
     }
 
-    @Override
     public boolean equals(Object object) {
-        if (this == object) {
-            return true;
-        } else {
-            return object instanceof ByteArrayTag && Arrays.equals(this.data, ((ByteArrayTag)object).data);
-        }
+        return this == object ? true : object instanceof ByteArrayTag && Arrays.equals(this.data, ((ByteArrayTag) object).data);
     }
 
-    @Override
     public int hashCode() {
         return Arrays.hashCode(this.data);
     }
@@ -115,32 +117,29 @@
         return this.data;
     }
 
-    @Override
     public int size() {
         return this.data.length;
     }
 
-    @Override
     public ByteTag get(int i) {
         return ByteTag.valueOf(this.data[i]);
     }
 
-    @Override
-    public ByteTag set(int i, ByteTag byteTag) {
-        byte b = this.data[i];
-        this.data[i] = byteTag.getAsByte();
-        return ByteTag.valueOf(b);
+    public ByteTag set(int i, ByteTag nbttagbyte) {
+        byte b0 = this.data[i];
+
+        this.data[i] = nbttagbyte.getAsByte();
+        return ByteTag.valueOf(b0);
     }
 
-    @Override
-    public void add(int i, ByteTag byteTag) {
-        this.data = ArrayUtils.add(this.data, i, byteTag.getAsByte());
+    public void add(int i, ByteTag nbttagbyte) {
+        this.data = ArrayUtils.add(this.data, i, nbttagbyte.getAsByte());
     }
 
     @Override
     public boolean setTag(int index, Tag element) {
         if (element instanceof NumericTag) {
-            this.data[index] = ((NumericTag)element).getAsByte();
+            this.data[index] = ((NumericTag) element).getAsByte();
             return true;
         } else {
             return false;
@@ -150,7 +149,7 @@
     @Override
     public boolean addTag(int index, Tag element) {
         if (element instanceof NumericTag) {
-            this.data = ArrayUtils.add(this.data, index, ((NumericTag)element).getAsByte());
+            this.data = ArrayUtils.add(this.data, index, ((NumericTag) element).getAsByte());
             return true;
         } else {
             return false;
@@ -159,9 +158,10 @@
 
     @Override
     public ByteTag remove(int i) {
-        byte b = this.data[i];
+        byte b0 = this.data[i];
+
         this.data = ArrayUtils.remove(this.data, i);
-        return ByteTag.valueOf(b);
+        return ByteTag.valueOf(b0);
     }
 
     @Override
@@ -169,7 +169,6 @@
         return 1;
     }
 
-    @Override
     public void clear() {
         this.data = new byte[0];
     }
