--- a/net/minecraft/world/level/block/RootedDirtBlock.java
+++ b/net/minecraft/world/level/block/RootedDirtBlock.java
@@ -9,6 +9,7 @@
 import net.minecraft.world.level.block.state.BlockState;
 
 public class RootedDirtBlock extends Block implements BonemealableBlock {
+
     public RootedDirtBlock(BlockBehaviour.Properties settings) {
         super(settings);
     }
@@ -25,6 +26,6 @@
 
     @Override
     public void performBonemeal(ServerLevel world, RandomSource random, BlockPos pos, BlockState state) {
-        world.setBlockAndUpdate(pos.below(), Blocks.HANGING_ROOTS.defaultBlockState());
+        org.bukkit.craftbukkit.v1_19_R1.event.CraftEventFactory.handleBlockSpreadEvent(world, pos, pos.below(), Blocks.HANGING_ROOTS.defaultBlockState()); // CraftBukkit
     }
 }
