--- a/net/minecraft/world/entity/projectile/SpectralArrow.java
+++ b/net/minecraft/world/entity/projectile/SpectralArrow.java
@@ -11,6 +11,7 @@
 import net.minecraft.world.level.Level;
 
 public class SpectralArrow extends AbstractArrow {
+
     public int duration = 200;
 
     public SpectralArrow(EntityType<? extends SpectralArrow> type, Level world) {
@@ -42,8 +43,9 @@
     @Override
     protected void doPostHurtEffects(LivingEntity target) {
         super.doPostHurtEffects(target);
-        MobEffectInstance mobEffectInstance = new MobEffectInstance(MobEffects.GLOWING, this.duration, 0);
-        target.addEffect(mobEffectInstance, this.getEffectSource());
+        MobEffectInstance mobeffect = new MobEffectInstance(MobEffects.GLOWING, this.duration, 0);
+
+        target.addEffect(mobeffect, this.getEffectSource(), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ARROW); // CraftBukkit
     }
 
     @Override
