--- a/net/minecraft/world/level/material/FluidState.java
+++ b/net/minecraft/world/level/material/FluidState.java
@@ -26,8 +26,12 @@
     public static final int AMOUNT_MAX = 9;
     public static final int AMOUNT_FULL = 8;
 
+    // Paper start
+    protected final boolean isEmpty;
+    // Paper end
     public FluidState(Fluid fluid, ImmutableMap<Property<?>, Comparable<?>> propertiesMap, MapCodec<FluidState> codec) {
         super(fluid, propertiesMap, codec);
+        this.isEmpty = fluid.isEmpty(); // Paper - moved from isEmpty()
     }
 
     public Fluid getType() {
@@ -43,7 +47,7 @@
     }
 
     public boolean isEmpty() {
-        return this.getType().isEmpty();
+        return this.isEmpty; // Paper - moved into constructor
     }
 
     public float getHeight(BlockGetter world, BlockPos pos) {
